### ISSUE TITLE
investigate how to speed up prebuilds on gitpod repo

### ISSUE THREAD

**Issue by akosyakov** [gitpod-io/gitpod#9874]
### Bug description

Right now there is a suspicion that gradle projects and Intellij indexing of them takes a lot of time. We should confirm whether it is true by profiling prebuilds somehow and understand why incremental prebuilds simply does not reuse dependencies and caches from last prebuild. We change these projects very rare.

### Steps to reproduce

Run a prebuild on this repo.

**Comment by felladrin** [comment]
I've noticed `components/ide/jetbrains/gateway-plugin:publish` is being triggered every time due to this `publishToJBMarketplace: true` in [WORKSPACE.yaml](https://github.com/gitpod-io/gitpod/blob/a766836727e7e9e0340332947d14a2a32b8dde8d/WORKSPACE.yaml). (Take [this Werft build log](https://werft.gitpod-dev.com/job/gitpod-build-main.3179/raw) as example)

Using GitHub Code Search you can see that `publishToJBMarketplace` is aways `true` which will cause Gateway Plugin to always rebuild:
- https://cs.github.com/gitpod-io/gitpod?q=jbMarketplacePublishTrigger
- https://cs.github.com/gitpod-io/gitpod?q=publishToJBMarketplace

<img width="1183" alt="image" src="https://user-images.githubusercontent.com/418083/167485983-96c86809-7ae9-41e6-b799-55bddb1f3041.png">

<img width="1239" alt="image" src="https://user-images.githubusercontent.com/418083/167486186-e9967823-9ede-4a5a-90d6-560fc012b2d4.png">

**Comment by Explosion-Scratch** [comment]
Yeah prebuilds are taking a really long time for me (+10min). It would be nice to simply not wait for the prebuild but instead see the progress of it in the editor via terminal

**Comment by akosyakov** [comment]
@felladrin Is it the same effect in prebuilds?

**Comment by mustard-mh** [comment]
As you can see in this prebuild job https://hw-gradle-research.preview.gitpod-dev.com/projects/gitpod/c4d70b3b-fc76-4df4-8312-b2eed367b452

|Task|StartTime|EndTime| Cost|
|:-:|:-:|:-:|:-:|
|gradle|Tue 10 May 2022 03:43:35 PM UTC|Tue 10 May 2022 03:45:04 PM UTC| 1:30|
|yarn & build|Tue 10 May 2022 03:43:16 PM UTC|Tue 10 May 2022 03:48:07 PM UTC|5:00|
|go|Tue 10 May 2022 03:44:44 PM UTC|Tue 10 May 2022 03:44:57 PM UTC|0:15|

incremental prebuilds works well in gradle, but not yarn & build

see also [incremental-prebuilds-result-log.zip](https://github.com/gitpod-io/gitpod/files/8663845/inc-pb-result.zip)

you need to search with words below to see job after pre-prebuild loaded
```
♻️ Re-running task as an incremental workspace prebuild
```

**Comment by mustard-mh** [comment]
incremental prebuilds will download pre-prebuild (`base`) result first, and this step will cost a lot of time too => prebuild in `US cluster` but `base` storage in `EU cluster`

**Comment by mustard-mh** [comment]
The road will be like this

1. Trigger incremental prebuild in `US cluster`
2. Workspace download `base` which storage in `EU cluster`
3. After prebuild, upload result to `EU cluster` ( in a `US cluster` workspace 😅
4. Start a workspace, to make user access in IDE, which will download prebuild result in `EU cluster` to `US cluster` workspace

**Comment by mustard-mh** [comment]
How to improve it?

- we need `node_modules` for code intelligent, `yarn build` can only work in component like `gitpod-protocol`. component like `dashboard` **don't need** to build first
- `gradle` should not be inside prebuild, it will gen folder about 7G, and every start will download it, and every stop will store it, it cost a lot.
- `gradle` can be gen after ws start, it'll only cost around 5 mins for all.

improve improvements above can reduce size of prebuild, so it can reduce the time for back and forth in download and upload base prebuild to reduce time of incremental prebuilds.

**Comment by mustard-mh** [comment]
incremental prebuilds got some conditions, you can check them out here https://github.com/gitpod-io/gitpod/blob/81a79af4ec3f9a5de53f40370ab455bf8a054c85/components/server/ee/src/workspace/workspace-factory.ts#L194

**Comment by mustard-mh** [comment]
This result **maybe** caused by the trigger of prebuild - Github, is in US🙈

**Comment by akosyakov** [comment]
@felladrin It will imply that we need to wait 7 mins to work on backend plugin each time? 5 mins full gralde build + 2 mins of intellij indexing?

**Comment by akosyakov** [comment]
> we need node_modules for code intelligent, yarn build can only run in component like gitpod-protocol. component like dashboard don't need to build first

@mustard-mh I did not entirely understand this point. But please file a new issue for that and add to webapp inbox. cc @jankeromnes @geropl

**Comment by csweichel** [comment]
> @felladrin It will imply that we need to wait 7 mins to work on backend plugin each time? 5 mins full gralde build + 2 mins of intellij indexing? I guess we indeed need to move to another rep and enable prebuilds there. Although moving to another repo does not solve the problem of long download/upload times? Pulling 10G should still be faster than 7mins? Let's move to command for now, and start looking in new setup for JB development or a discussion about running prebuild in the same cluster where they are stored?

The trade-off that we're looking at today is: "spend 10 minutes for everyone outside of the EU region to download the prebuild + egress cost" vs "spend 7 minutes for anyone who wants to work on JB, without the egress cost". 

Hence, removing the gradle command for now makes 💯 sense. Moving to another repo - as you pointed out - does not solve the issue.
This prebuild performance problem will hopefully be solved with the incoming PVC work. PVC prebuilds are still a few weeks out, but we're not that far away. And, this will be a great test-case for their performance.

That's to say, remove gradle now, wait for PVC and put it back in then.

**Comment by akosyakov** [comment]
ok @mustard-mh could you make a PR about moving command and we pause on it for now

**Comment by mustard-mh** [comment]
moving to command doesn't seem to be enough, because it would make everyone who opens the gitpod repo generate 6GB of gradle cache, which is also costly to store on GS, and we advocate the concept of workspace use it up, so a lot of useless storage would be wasted, so I think we So I think we can add a command like `read -p "Press enter to continue"` to make the gradle not run by default, and the command will be activated explicitly only after we press the key, This is because Jetbrains developers make up a small percentage of the Gitpoder WDYT @akosyakov

**Comment by akosyakov** [comment]
@mustard-mh Could we introduce a env variable to control it? I can enable it for me always with settings then.

**Comment by mustard-mh** [comment]
@akosyakov That's fine

**Comment by akosyakov** [comment]
Let's keep this issue opened and mark tech debt till the situation with prebuild is storage is not improved. Then we should put everythin back and try agin.

**Comment by felladrin** [comment]
Now that #9927 has been merged, I've noticed that even if I don't press enter to start the Java task, I still have intellisense for kotlin files in `backend-plugin` and `gateway-plugin` when I open Gitpod Repo on IntelliJ IDEA. 🤔 Makes me wonder if we actually need to run that task.

- #9927

**Comment by akosyakov** [comment]
@felladrin How fast is your dev loop when you try changes? gradle build is not relevant for IntelliJ, but for testing

**Comment by mustard-mh** [comment]
related https://github.com/gitpod-io/gitpod/issues/6650

**Comment by felladrin** [comment]
> @felladrin How fast is your dev loop when you try changes?

For backend-plugin I open the repo on Gitpod IntelliJ IDEA and run `launch-dev-server.sh` and manually restart the shell script after changing the code.

For the gateway-plugin I use local IntellJ IDEA, so it's always relying on what has been previously cached here by `./gradlew runIde`.

(Not sure if this answers the question. But I can say it's already faster to develop JetBrains plugins, compared to 3 months ago.)

> gradle build is not relevant for IntelliJ, but for testing

Can you tell more about it? Is it about `./gradlew check`?

**Comment by akosyakov** [comment]
No, I mean that `./gradlew runIde` will run build and it was already executed then you don't need to wait for it.
